imap/index.c

```diff
@@ -5363,161 +5319,96 @@ extractor_timeout(struct protstream *s __attribute__((unused)),
-        syslog(LOG_DEBUG, "extractor_connect: %sconnect to %s",
-               be ? "re" : "", hostname);
-
-        be = ext->be = backend_connect(be, hostname,
-                                       &http, NULL, NULL, NULL, -1);
-
-        if (!be) {
-            syslog(LOG_ERR, "extract_connect: failed to connect to %s",
-                   hostname);
-            ext->failed++;
-            r = IMAP_IOERROR;
-        }
-        else if (!ctx) {
-            ctx = be->context = xzmalloc(sizeof(struct backend_ctx));
-            ctx->path = xstrdup(path);
-
-            if (ext->clientin) {
-                /* add a default timeout */
-                be->clientin = ext->clientin;
-                be->timeout = prot_addwaitevent(ext->clientin,
-                                                now + IDLE_TIMEOUT,
-                                                extractor_timeout, be);
-            }
-        }
+    if (ext->clientin) {
+        /* add a default timeout */
+        be->clientin = ext->clientin;
+        be->timeout = prot_addwaitevent(ext->clientin,
+                                        now + IDLE_TIMEOUT,
+                                        extractor_timeout, ctx);
     }
 
-    if (be->timeout && ctx->timeout) be->timeout->mark = now + ctx->timeout;
-
-    buf_free(&buf);
-
-    return r;
+    return 0;
 }
 
-
 static int extract_attachment(const char *type, const char *subtype,
                               const struct param *type_params,
                               const struct buf *data, int encoding,
                               const struct message_guid *content_guid,
                               struct getsearchtext_rock *str)
 {
-    const char *exturl =
-         config_getstring(IMAPOPT_SEARCH_ATTACHMENT_EXTRACTOR_URL);
     struct backend *be;
-    struct backend_ctx *ctx;
     struct buf buf = BUF_INITIALIZER;
     hdrcache_t hdrs = NULL;
     struct body_t body = { 0, 0, 0, 0, 0, BUF_INITIALIZER };
     const char *guidstr, *errstr = NULL;
     size_t hostlen;
+    const char **hdr, *p;
     int r = 0;
 
-    if (!str->ext) {
-        if (!exturl) return 0;
-        if (!index_text_extractor) {
-            /* This is a legitimate case for sieve and lmtpd (so we don't need
-             * to spam the logs! */
-            syslog(LOG_DEBUG, "%s: ignoring uninitialized extractor for url %s",
-                    __func__, exturl);
-            return 0;
-        }
-
-        /* cache the connection for all the attachments in this message, but only
-         * if the connection succeeds */
-        r = extractor_connect(exturl, index_text_extractor);
-        if (r) return r;
-        str->ext = index_text_extractor;
+    if (!index_text_extractor) {
+        /* This is a legitimate case for sieve and lmtpd (so we don't need
+         * to spam the logs! */
+        syslog(LOG_DEBUG, "%s: ignoring uninitialized extractor",
+                __func__);
+        return 0;
```

Comment:
Since all the exturl checking is up-front, we just check if there's an extractor configured at all by checking for the global.